mlfinlab/portfolio_optimization/nco.py

```diff
@@ -0,0 +1,362 @@
+# pylint: disable=missing-module-docstring
+import numpy as np
+import pandas as pd
+from sklearn.covariance import LedoitWolf
+from sklearn.cluster import KMeans
+from sklearn.metrics import silhouette_samples
+from scipy.linalg import block_diag
+from mlfinlab.portfolio_optimization.risk_estimators import RiskEstimators
+
+
+class NCO:
+    """
+    This class implements the Nested Clustered Optimization (NCO) algorithm. It is reproduced with
+    modification from the following paper: `Marcos Lopez de Prado “A Robust Estimator of the Efficient Frontier”,
+    (2019). <https://papers.ssrn.com/abstract_id=3469961>`_.
+    """
+
+    def __init__(self):
+        """
+        Initialize
+        """
+
+    @staticmethod
+    def _simulate_covariance(mu_vector, cov_matrix, num_obs, lw_shrinkage=False):
+        """
+        Derives an empirical vector of means and an empirical covariance matrix.
+
+        Based on the set of true means vector and covariance matrix of X distributions,
+        the function generates num_obs observations for every X.
+        Based on these observations simulated vector of means and the simulated covariance
+        matrix are obtained.
+
+        :param mu_vector: (np.array) True means vector for X distributions
+        :param cov_matrix: (np.array) True covariance matrix for X distributions
+        :param num_obs: (int) Number of observations to draw for every X
+        :param lw_shrinkage: (bool) Flag to apply Ledoit-Wolf shrinkage to X
+        :return: (np.array, np.array) Empirical means vector, empirical covariance matrix
+
+        """
+
+        # Generating a matrix of num_obs observations for X distributions
+        observations = np.random.multivariate_normal(mu_vector.flatten(), cov_matrix, size=num_obs)
+
+        # Empirical means vector calculation
+        mu_simulated = observations.mean(axis=0).reshape(-1, 1)
+
+        if lw_shrinkage:  # If applying Ledoit-Wolf shrinkage
+            cov_simulated = LedoitWolf().fit(observations).covariance_
+
+        else:  # Simple empirical covariance matrix
+            cov_simulated = np.cov(observations, rowvar=False)
+
+        return mu_simulated, cov_simulated
+
+    @staticmethod
+    def _cluster_kmeans_base(corr, max_num_clusters=None, n_init=10):
+        """
+        Finding the optimal partition of clusters using K-Means algorithm.
+
+        For the fit of K-Means algorithm a matrix of distances based on the correlation matrix is used.
+        The algorithm iterates through n_init - time K-Means can run with different seeds and
```

Comment:
Change this line to - "the algorithm iterates n_init number of times and initialises K-Means with different seeds and max_number_of_clusters"

Reply:
Done